Added a config to export the library as a cmake target

hellozee opened the pull request:
This exports the library as a cmake package with `sqlpp11::sqlpp-mysql` target. Don't know if it is necessary or not, but I needed it to link against my program when installed this library from vcpkg. I have kept it under the `sqlpp11` namespace, since package name is pretty long.

hellozee committed:
Added a config to export the library as a cmake target

rbock commented on src/CMakeLists.txt:
```diff
@@ -6,3 +6,36 @@ add_library(sqlpp-mysql
+
+write_basic_package_version_file(
+    "${CMAKE_CURRENT_BINARY_DIR}/cmake/sqlpp11-connector-mysqlConfigVersion.cmake"
+        VERSION 0.29
```
This has to be automated. Otherwise it will be outdated whenever the next release comes :-)

hellozee replied:
Sorry was a quick hack. Used the PROJECT_VERSION variable instead now, cause I can't think of anything better.

hellozee replied:
Also ran cmake-format cause there was inconsistencies in the code that I copied from multiple places.

hellozee committed:
Made use of PROJECT_VERSION instead of hardcoding the number

hellozee committed:
Ran cmake-format over src/CMakeLists.txt

rbock commented on CMakeLists.txt:
```diff
@@ -24,7 +24,7 @@
 cmake_minimum_required(VERSION 3.2)
 include(CheckCXXSymbolExists)
-project (sqlpp11-connector-mysql)
+project (sqlpp11-connector-mysql VERSION 0.29)
```
The version is defined by the tag in git. Maintaining it separately in a file is not an option.

I know myself: I will forget it starting in 0.30

hellozee replied:
Fair enough, https://github.com/rbock/sqlpp11/blob/develop/CMakeLists.txt#L28 😛 
I will check if there is something, probably some cmake module which can read the version from tag and set that up during compile time.

hellozee committed:
pick the version from the current tag

hellozee commented:
Sorry for being late, 😸

rbock commented:
Thanks for the update. I appreciate the effort you put into this.

However, this now has hundreds of lines of code for the equivalent of `git tag | sort -V | tail -n1`.

Your initial statement was:

> Don't know if it is necessary or not, but I needed it to link against my program when installed this library from vcpkg.

I have no relevant knowledge of vcpkg and my cmake skills are rather limited. I cannot imagine that anything like this is necessary to make them work together. Please do more research. There has to be a better way.

Thanks!

hellozee committed:
removed the newly introduced modules

hellozee commented:
> However, this now has hundreds of lines of code for the equivalent of `git tag | sort -V | tail -n1`.

Sorry, that was me being lazy, I copied a whole module where I needed just a single function, removed that and replaced that with a small command.

> Your initial statement was:
> 
> > Don't know if it is necessary or not, but I needed it to link against my program when installed this library from vcpkg.
> 
> I have no relevant knowledge of vcpkg and my cmake skills are rather limited. I cannot imagine that anything like this is necessary to make them work together. Please do more research. There has to be a better way.

I can't link against the connector since no targets were exported corresponding to it. So unless the libraries and headers exist in the system path, it would be a hassle to use it. That is what I think, though I could be definitely wrong.

rbock commented on CMakeLists.txt:
```diff
@@ -33,6 +33,15 @@ set(CMAKE_CXX_STANDARD 11)
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+
+execute_process(
+	COMMAND git describe --tags
```
This would break if for instance `git` is not in the path or (more likely) someone downloaded the zip without `.git` directory.

In that case `TAG` would be unset, I guess? Is that a problem?

rbock commented on src/CMakeLists.txt:
```diff
@@ -1,8 +1,32 @@
+
+write_basic_package_version_file(
+  "${CMAKE_CURRENT_BINARY_DIR}/cmake/sqlpp11-connector-mysqlConfigVersion.cmake"
+  VERSION ${TAG}
```
If `TAG` is undefined, I believe this will be an error.

We call `write_basic_package_version_file` only in case that `TAG` is actually defined...

hellozee committed:
sets tag if the directory is not a git repo or git command is not available

hellozee commented:
Put a check for the TAG variable if not set, I am out of ideas on how to tackle this if there is not .git or `git` isnt available in the system, :)

rbock commented:
Thanks!

@Leon0402 and @GCarneiroA My cmake foo is insufficient to figure out if this would conflict with #65?

Leon0402 commented on src/CMakeLists.txt:
```diff
@@ -1,8 +1,32 @@
+
+install(
+  FILES
+    "../cmake/sqlpp11-connector-mysqlConfig.cmake"
```
Why did you copy the config file in the binary directory in line 17, if you now use the path to the cmake directory directly? 

Also instead of using "../cmake", you could also use "${PROJECT_SOURCE_DIR}/cmake/", which I personally like more. But that might be preference.

Leon0402 commented on src/CMakeLists.txt:
```diff
@@ -1,8 +1,32 @@
+  "${CMAKE_CURRENT_BINARY_DIR}/cmake/sqlpp11-connector-mysqlConfig.cmake"
+  COPYONLY)
+
+set(ConfigPackageLocation lib/cmake/sqlpp11-connector-mysql)
```
To be compatible with all major distributions / OS, you should rather not hardcode the destinations path here. Instead cmake can figure out the correct directory: 
```
include(GNUInstallDirs)
set(ConfigPackageLocation ${CMAKE_INSTALL_LIBDIR}/cmake/sqlpp11-connector-mysql)
```

Leon0402 commented on src/CMakeLists.txt:
```diff
@@ -1,8 +1,32 @@
+add_library(sqlpp-mysql connection.cpp prepared_statement.cpp char_result.cpp
+                        bind_result.cpp detail/connection_handle.cpp)
 
 install(TARGETS sqlpp-mysql DESTINATION lib)
```
This can be removed I guess

Leon0402 commented on src/CMakeLists.txt:
```diff
@@ -1,8 +1,32 @@
+  VERSION ${TAG}
+  COMPATIBILITY AnyNewerVersion)
+
+export(
```
This shouldn't be necessary, it is done by install(EXPORT ...) in line 23

Leon0402 commented on src/CMakeLists.txt:
```diff
@@ -1,8 +1,32 @@
+  FILES
+    "../cmake/sqlpp11-connector-mysqlConfig.cmake"
+    "${CMAKE_CURRENT_BINARY_DIR}/cmake/sqlpp11-connector-mysqlConfigVersion.cmake"
+  DESTINATION ${ConfigPackageLocation})
```
Actually nothing here, but I can't comment on code you haven't changed (Why Github?). In the top level cmake file there is the installation command for the header files. You might wanna move them in here, to have everything installation related in one place

Leon0402 commented on cmake/sqlpp11-connector-mysqlConfig.cmake:
```diff
@@ -0,0 +1,6 @@
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}")
```
Seems to be not needed

Leon0402 commented:
> Thanks!
> 
> @Leon0402 and @GCarneiroA My cmake foo is insufficient to figure out if this would conflict with #65?

It should be alright as far as I can see. 

But @GCarneiroA made some good changes over there, which this patch could benefit from. He added target specific commands like `target_include_directories` with the appropriate visibility (most often PUBLIC here). So everyone that links against this library will have set the include directories, libraries set automatically. 
There is special handling  for Installation though. The path of the headers will change. Therefore you would usually add some logic to make target_include_directories contain relative paths when building, but pathes to the installation folder when installing. It might look like this: 
```
target_include_directories(mysql-connector 
   INTERFACE
       $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include>
       $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
)
```
The INSTALL_INTERFACE part can be removed, when  using `INCLUDES DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}` on the `install(TARGETS ...)` command. I use it normally, because you can't do anything wrong here :) 

As far as I can see the @GCarneiroA took this into account in his patch, although the INSTALL_INTERFACE part isn't correct I think. 

So I think it might be better to merge the other PR first and then rebase here. But I'm not sure if  @GCarneiroA  is still working at the other PR, it seems stale.

rbock commented:
The mysql-connector now lives in the sqlpp11 repo.

Closing this before archiving. Please re-open in sqlpp11, if this is still relevant.

Thanks!
